### PR TITLE
nixos/mastodon: fix init db on remote postgresql

### PR DESCRIPTION
**Pull request by Izorkin**
###### Motivation for this change
Fix script to init db on remote postgresql.

cc @erictapen

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Izorkin** [comment]
cc @erictapen

**Comment by Izorkin** [comment]
cc @erictapen

**Comment by aanderse** [comment]
I don't think NixOS as a distribution should do things like this :man_shrugging:

**Comment by Izorkin** [comment]
@aanderse the service shouldn't create a database remotely? Without these changes, the mastodon-init-db service does not work for me.

**Comment by Izorkin** [comment]
cc @erictapen

**Comment by erictapen** [comment]
Hey @Izorkin, sorry for not getting back for such a long time. I'm seeing your pings and I'm aware of this PR (and the other ones you created regarding mastodon). The reason I'm hesitant in approving these is that I'm currently unable to test the changes.

I have no setup with a remote postgresql db and I don't really have time to set one up.

The best way to test these changes would be to have a proper NixOS integration test for Mastodon, that tests both an instance with local redis,postgresql,postfix and one with remote services. But writing that kind of test takes time, which is sparse for me.

I can see that this is an unsatisfiying situation, but I'm afraid I don't really see an easy way out here. In the meantime I'd ask you to refrain from pinging me further on this issue. I'm aware of it and I'll get to it as soon as I have the capacity to test it properly. That being said I'd have nothing against others to test and review this change. It's just that I'm not confident enough to merge this without breaking other people's setups.

**Comment by Izorkin** [comment]
I tested these changes on test virtual machines in various combinations.

**Comment by erictapen** [comment]
That's cool, but the point of the review process is to have an additional person look over it. Find one approving it and I'll happily merge it.

**Comment by Izorkin** [comment]
cc @erictapen

**Comment by Izorkin** [comment]
@GrahamcOfBorg test mastodon-extdb

**Comment by Izorkin** [comment]
@GrahamcOfBorg test mastodon-extdb

File: nixos/tests/all-tests.nix [hunk 1]
```diff
@@ -287,6 +287,7 @@ in
   man = handleTest ./man.nix {};
   mariadb-galera = handleTest ./mysql/mariadb-galera.nix {};
   mastodon = handleTestOn ["x86_64-linux" "i686-linux" "aarch64-linux"] ./web-apps/mastodon.nix {};
+  mastodon-extdb = handleTestOn ["x86_64-linux" "i686-linux" "aarch64-linux"] ./web-apps/mastodon-extdb.nix {};
```
Comment (erictapen) [inline nixos/tests/all-tests.nix:290]: This shouldn't be a separate test, but rather an addition to the one we have. Too much duplicate code otherwise.

We should abstract over running a test against a mastodon instance, e.g. by introducing a function `make-mastodon-test`, that accepts `nodes`. See `nixos/tests/postgresql.nix` for inspiration, even though there it takes the postgresql version as an argument.

Also perspectively, but that is outside the scope of this PR, I think it would be cool if we'd test that all components that one can run externally (postgresql, nginx, redis, smtp) are also tested. What do you think about that?

Comment (Izorkin) [reply nixos/tests/all-tests.nix:290]: > This shouldn't be a separate test, but rather an addition to the one we have. Too much duplicate code otherwise.
> 
> We should abstract over running a test against a mastodon instance, e.g. by introducing a function `make-mastodon-test`, that accepts `nodes`. See `nixos/tests/postgresql.nix` for inspiration, even though there it takes the postgresql version as an argument.

Now I don't know how to implement such a test.



> Also perspectively, but that is outside the scope of this PR, I think it would be cool if we'd test that all components that one can run externally (postgresql, nginx, redis, smtp) are also tested. What do you think about that?

Running nginx from other server doesn't make sense

By default, only running `mastodon` test.
Test `mastodon-extdb` serves as an additional check.
For example - tests of nginx, systemd, mysql, run independently of each other.

Comment (erictapen) [reply nixos/tests/all-tests.nix:290]: Both would be fine for me!

Comment (Izorkin) [reply nixos/tests/all-tests.nix:290]: Now tested this option, the presence of nginx on the server with mastodon is mandatory. I don't see the point of separating nginx in the test.

Comment (Izorkin) [reply nixos/tests/all-tests.nix:290]: No full checking this variant.

Comment (Izorkin) [reply nixos/tests/all-tests.nix]: Created!

**Comment by Izorkin** [comment]
@GrahamcOfBorg test mastodon-ext

**Comment by Izorkin** [comment]
@GrahamcOfBorg test mastodon-ext

**Comment by Izorkin** [comment]
cc @erictapen

**Comment by erictapen** [comment]
@Izorkin Please, and I think I'm repeating myself here, please stop reminding me about outstanding issues without adding any further info on the topic. I'm aware of this and your other PRs and I have them saved in my todo list. If I'm not responding to an issue for a while, it usually means that I haven't got enough time to deal with it or that I don't prioritise it high enough to spend my resources on it at the moment.

I'm currently the only Mastodon maintainer with commit bit. That is a frustrating situation, but I'm afraid you have to accept it for the moment. If you want to have your stuff merged either ask somebody else with commit access or be patient with me. But pinging me [twice a](https://github.com/NixOS/nixpkgs/pull/165036#issuecomment-1083268369) [day](https://github.com/NixOS/nixpkgs/pull/117126#issuecomment-1082353503) just for the purpose of reminding me of this issue is rude tbh.

Had to get this off my chest for a while now. I can see that it is frustrating to put work into something and then not have it merged for a while. But please keep in mind, that I'm doing this in my free time as well. And your frequent "cc @erictapen" pings make this maintainership a lot less fun for me.

That being said, I value the many contributions you brought to the package and module. I hope we can figure out a mode of collaboration that works out for both of us.

**Comment by Izorkin** [comment]
Hope there will be more maintainers :)
I wanted to close the current task quickly.

**Comment by Izorkin** [comment]
Excuse me for being overly persistent.

**Comment by erictapen** [comment]
@Izorkin Apology accepted, I appreciate it. Let's continue to build cool stuff together :)

File: nixos/modules/services/web-apps/mastodon.nix [hunk 2]
```diff
@@ -35,6 +36,7 @@ let
 
     TRUSTED_PROXY_IP = cfg.trustedProxy;
   }
+  // (if (cfg.database.host != "/run/postgresql" && cfg.database.port != null) then { DB_PORT = toString(cfg.database.port); } else {})
```
Comment (erictapen) [inline nixos/modules/services/web-apps/mastodon.nix:39]:
```suggestion
  // (if (cfg.database.host != "/run/postgresql" && cfg.database.port != null) then { DB_PORT = toString cfg.database.port; } else {})
```

File: nixos/modules/services/web-apps/mastodon.nix [hunk 3]
```diff
@@ -330,7 +337,7 @@ in {
 
         passwordFile = lib.mkOption {
           type = lib.types.nullOr lib.types.path;
-          default = "/var/lib/mastodon/secrets/db-password";
+          default = null;
```
Comment (erictapen) [inline nixos/modules/services/web-apps/mastodon.nix:340]: I think this should be mentioned in release notes, as we change a default here.

File: nixos/modules/services/web-apps/mastodon.nix [hunk 4]
```diff
@@ -477,17 +497,31 @@ in {
...
+      script = lib.optionalString (!databaseActuallyCreateLocally) ''
+        umask 077
+
+        export PGPASSFILE
+        PGPASSFILE=$(mktemp)
+        cat > $PGPASSFILE <<EOF
+        ${cfg.database.host}:${toString cfg.database.port}:${cfg.database.name}:${cfg.database.user}:$(cat ${cfg.database.passwordFile})
+        EOF
+
+        if [ $(psql --host ${cfg.database.host} --user ${cfg.database.user} ${cfg.database.name} -c \
```
Comment (erictapen) [inline nixos/modules/services/web-apps/mastodon.nix:509]: I think `--host` and `--user` should be set via the PGHOST and PGUSER environment variables, which should be defined in `systemd.services.mastodon-init-db.environment`. That way we don't have to generate that much bash code in Nix.

**Review by erictapen — CHANGES_REQUESTED** [review]

File: nixos/modules/services/web-apps/mastodon.nix (hunk 3 above)
Comment (Izorkin) [reply nixos/modules/services/web-apps/mastodon.nix:342]: In this case a warning will appear:
```
error:
       Failed assertions:
       - <option>services.mastodon.database.passwordFile</option> needs to be set if
         <option>services.mastodon.database.createLocally</option> is not enabled.
(use '--show-trace' to show detailed location information)

```

**Comment by Izorkin** [comment]
@erictapen thanks, updated PR.

**Comment by Izorkin** [comment]
@GrahamcOfBorg test mastodon-ext

**Comment by Izorkin** [comment]
@GrahamcOfBorg test mastodon

File: nixos/modules/services/web-apps/mastodon.nix (hunk 3 above)
Comment (erictapen) [reply nixos/modules/services/web-apps/mastodon.nix:342]: Right, makes sense! No need for release notes then.

**Comment by Izorkin** [comment]
@GrahamcOfBorg test mastodon-ext

**Comment by Izorkin** [comment]
@GrahamcOfBorg test mastodon mastodon-ext

**Comment by Izorkin** [comment]
@erictapen please check and review PR.

**Comment by turion** [comment]
Can you rebase?

**Comment by Izorkin** [comment]
Rebased PR.

File: nixos/tests/all-tests.nix [hunk 5]
```diff
@@ -361,6 +361,7 @@ in {
   man = handleTest ./man.nix {};
   mariadb-galera = handleTest ./mysql/mariadb-galera.nix {};
   mastodon = handleTestOn ["x86_64-linux" "i686-linux" "aarch64-linux"] ./web-apps/mastodon.nix {};
+  mastodon-ext = handleTestOn ["x86_64-linux" "i686-linux" "aarch64-linux"] ./web-apps/mastodon-ext.nix {};
```
Comment (turion) [inline nixos/tests/all-tests.nix:364]: It would be nicer to put both mastodon tests in a directory and then use `discoverTests`

File: nixos/tests/web-apps/mastodon-ext.nix [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+        extraHosts = hosts;
+      };
+
+      security = {
+        pki.certificateFiles = [ "${test-certificates}/root_ca.crt" ];
+      };
+    };
+  };
+
+  testScript = ''
```
Comment (turion) [inline nixos/tests/web-apps/mastodon-ext.nix:200]: The test script is the same for both tests, right? Can you share it as a single file?

File: nixos/tests/all-tests.nix (hunk 5 above)
Comment (Izorkin) [reply nixos/tests/all-tests.nix:364]: How to combine `handleTestOn` + `discoverTests` ?

Comment (Izorkin) [reply nixos/tests/all-tests.nix:364]: 362|   mariadb-galera = handleTest ./mysql/mariadb-galera.nix {};
          363|   mastodon = discoverTests (import ./web-apps/mastodon {});
             |                             ^
          364|   matomo = handleTest ./matomo.nix {};

File: nixos/tests/web-apps/mastodon-ext.nix (hunk 6 above)
Comment (Izorkin) [reply nixos/tests/web-apps/mastodon-ext.nix:200]: The test script is slightly different:
```diff
--- test1       2022-11-15 18:38:19.452250259 +0000
+++ test2       2022-11-15 18:38:39.520217996 +0000
@@ -4,9 +4,12 @@
     ca.wait_for_unit("step-ca.service")
     ca.wait_for_open_port(8443)

-    server.wait_for_unit("acme-finished-mastodon.local.target")
+    database.wait_for_unit("postgresql.service")
+    database.wait_for_open_port(5432)
+
+    nginx.wait_for_unit("acme-finished-mastodon.local.target")
+
     server.wait_for_unit("redis-mastodon.service")
-    server.wait_for_unit("postgresql.service")
     server.wait_for_unit("mastodon-sidekiq.service")
     server.wait_for_unit("mastodon-streaming.service")
     server.wait_for_unit("mastodon-web.service")
@@ -45,4 +48,6 @@
     ca.shutdown()
     server.shutdown()
     client.shutdown()
+    database.shutdown()
+    nginx.shutdown()
   '';
```
How to implement in one file do not know

**Comment by Izorkin** [comment]
If I now delete the commits with tests, then it will be possible to merge the PR?

File: nixos/tests/all-tests.nix (hunk 5 above)
Comment (Izorkin) [reply nixos/tests/all-tests.nix:365]: @turion can we keep current variant? For example, all nginx tests are not combined into one test either.

Comment (Izorkin) [reply nixos/tests/all-tests.nix]: Fixed.

File: nixos/modules/services/web-apps/mastodon.nix [hunk 7]
```diff
@@ -465,7 +474,39 @@ in {
...
+      }
+      {
+        assertion = !databaseActuallyCreateLocally -> (cfg.database.host != "/run/postgresql");
+        message = ''
+          <option>services.mastodon.database.host</option> needs to be set if
+            <option>services.mastodon.database.createLocally</option> is not enabled.
+        '';
+      }
+      {
+        assertion = !databaseActuallyCreateLocally -> (cfg.database.passwordFile != null);
```
Comment (erictapen) [inline nixos/modules/services/web-apps/mastodon.nix:491]: There are other ways of authenticating against a remote PostgreSQL instance, e.g. via TLS client certificates. This assertion would break such a setup.

File: nixos/modules/services/web-apps/mastodon.nix [hunk 8]
```diff
@@ -511,7 +554,16 @@ in {
     };
 
     systemd.services.mastodon-init-db = lib.mkIf cfg.automaticMigrations {
-      script = ''
+      script = lib.optionalString (!databaseActuallyCreateLocally) ''
```
Comment (erictapen) [inline nixos/modules/services/web-apps/mastodon.nix:557]: I really don't like the additional shell code here. Can't we replace this with some additions to systemd.services.mastodon-init-db.environment?

File: nixos/modules/services/web-apps/mastodon.nix [hunk 9]
```diff
@@ -372,17 +379,19 @@ in {
         };
 
         user = lib.mkOption {
+          type = lib.types.nullOr lib.types.str;
```
Comment (erictapen) [inline nixos/modules/services/web-apps/mastodon.nix:382]: These additions don't have anything to do with the purpose of the PR, right? Please file them as another PR if that's correct.

File: nixos/modules/services/web-apps/mastodon.nix [hunk 10]
```diff
@@ -465,7 +474,39 @@ in {
...
+      }
+      {
+        assertion = !databaseActuallyCreateLocally -> (cfg.database.passwordFile != null);
+        message = ''
+          <option>services.mastodon.database.passwordFile</option> needs to be set if
+            <option>services.mastodon.database.createLocally</option> is not enabled.
+        '';
+      }
+      {
+        assertion = cfg.smtp.authenticate -> (cfg.smtp.user != null);
```
Comment (erictapen) [inline nixos/modules/services/web-apps/mastodon.nix:498]: Also smtp related?

**Review by erictapen — CHANGES_REQUESTED** [review]
@Izorkin I know this has been going on for over one and a half years and I'm sorry for that.

I have had multiple issues with this PR. In the beginning I told you that I don't feel comfortable merging code that isn't tested. Than I oftentimes had trouble getting to the gist of what you are actually doing here or what problem you are fixing.

But now that I have looked at the code for a about a few dozen times, I realize that maybe the whole idea of supporting remote PostgreSQL dbs was a mistake in that it adds too much complexity to the module than we are able to maintain at the moment.

Nevertheless I'm willing to give it another try. I'd be even okay if this goes forward without the test I wanted from you. We can add it at a later time, when the code duplication problems have been solved.

File: nixos/modules/services/web-apps/mastodon.nix (hunk 8 above)
Comment (Izorkin) [reply nixos/modules/services/web-apps/mastodon.nix:557]: I haven't found any other variants. And storing the database password in a variable is not secure.

File: nixos/modules/services/web-apps/mastodon.nix (hunk 9 above)
Comment (Izorkin) [reply nixos/modules/services/web-apps/mastodon.nix:382]: Okay, I'll move it to a another PR.

**Comment by Izorkin** [comment]
> Than I oftentimes had trouble getting to the gist of what you are actually doing here or what problem you are fixing.

I have a PostgreSQL database running on a physical server and mastodon running in a virtual machine. The current version of the mastodon service does not allow to work with the remote database.

> Nevertheless I'm willing to give it another try. I'd be even okay if this goes forward without the test I wanted from you. We can add it at a later time, when the code duplication problems have been solved.

In case you haven't noticed, I've already added a test with remote database support.

**Comment by erictapen** [comment]
> In case you haven't noticed, I've already added a test with remote database support.

I did notice! I was addressing your earlier question and @turion's concern about code duplication.

> If I now delete the commits with tests, then it will be possible to merge the PR?

**Comment by Izorkin** [comment]
> I did notice! I was addressing your earlier question and @turion's concern about code duplication.

I do not know how to implement this option :(
I can leave only one test to check service startup.

File: nixos/modules/services/web-apps/mastodon.nix (hunk 7 above)
Comment (Izorkin) [reply nixos/modules/services/web-apps/mastodon.nix:491]: I couldn't find certificate settings in `services.postgresql`.

**Comment by turion** [comment]
> I do not know how to implement this option :(

Hopefully I wasn't too pedantic with my request. If you want I can try and push a commit implementing this tomorrow.

**Comment by Izorkin** [comment]
> Hopefully I wasn't too pedantic with my request. If you want I can try and push a commit implementing this tomorrow.

Try it, maybe you can do it.

File: nixos/modules/services/web-apps/mastodon.nix (hunk 3 above)
Comment (turion) [reply nixos/modules/services/web-apps/mastodon.nix:342]: But people who have had a working setup before will have a broken one now without a simple means to fix it. They will not know what the previous location was and suddenly need to set it, it seems to me?

File: nixos/modules/services/web-apps/mastodon.nix [hunk 11]
```diff
@@ -511,7 +538,16 @@ in {
...
+      script = lib.optionalString (!databaseActuallyCreateLocally) ''
+        umask 077
+
+        export PGPASSFILE
+        PGPASSFILE=$(mktemp)
+        cat > $PGPASSFILE <<EOF
+        ${cfg.database.host}:${toString cfg.database.port}:${cfg.database.name}:${cfg.database.user}:$(cat ${cfg.database.passwordFile})
+        EOF
+
+      '' + ''
```
Comment (turion) [inline nixos/modules/services/web-apps/mastodon.nix:550]: Why add two strings instead of writing a long string?

Comment (Izorkin) [reply nixos/modules/services/web-apps/mastodon.nix:550]: You mean this code?
```
        cat > $PGPASSFILE <<EOF
        ${cfg.database.host}:${toString cfg.database.port}:${cfg.database.name}:${cfg.database.user}:$(cat ${cfg.database.passwordFile})
        EOF
```

**Comment by Izorkin** [comment]
Added release-notes.

File: nixos/modules/services/web-apps/mastodon.nix (hunk 11 above)
Comment (turion) [reply nixos/modules/services/web-apps/mastodon.nix:550]: No I mean why line 550 exists when all it does is concatenate two constant strings.

Comment (Izorkin) [reply nixos/modules/services/web-apps/mastodon.nix:550]: This is done in such a way that this code will only work when `databaseActuallyCreateLocally = false`:
```
        umask 077
...
```
And this code worked in both variants  `databaseActuallyCreateLocally`:
```
        if [ `psql ${cfg.database.name} -c \
...
```

Comment (turion) [reply nixos/modules/services/web-apps/mastodon.nix:550]: Ah yes thanks! My bad.

File: nixos/tests/web-apps/mastodon/default.nix [hunk 12]
```diff
@@ -0,0 +1,5 @@
+{ system ? builtins.currentSystem, handleTestOn }:
+{
+  mastodon-standart = handleTestOn [ "x86_64-linux" "i686-linux" "aarch64-linux" ] ./mastodon-std.nix { inherit system; };
```
Comment (turion) [inline nixos/tests/web-apps/mastodon/default.nix:3]:
```suggestion
  standard = handleTestOn [ "x86_64-linux" "i686-linux" "aarch64-linux" ] ./standard.nix { inherit system; };
```
This seems like a good solution to group the tests! Now since `mastodon` then already serves as a namespace in the attribute names and file names, we can simplify this, fix a typo and make it consistent.

File: nixos/tests/web-apps/mastodon/default.nix [hunk 13]
```diff
@@ -0,0 +1,5 @@
...
+  mastodon-extended = handleTestOn [ "x86_64-linux" "i686-linux" "aarch64-linux" ] ./mastodon-ext.nix { inherit system; };
```
Comment (turion) [inline nixos/tests/web-apps/mastodon/default.nix:4]:
```suggestion
let
  supportedSystems = [ "x86_64-linux" "i686-linux" "aarch64-linux" ];
{
  mastodon-standart = handleTestOn supportedSystems ./mastodon-std.nix { inherit system; };
  mastodon-extended = handleTestOn supportedSystems ./mastodon-ext.nix { inherit system; };
```

File: nixos/tests/web-apps/mastodon/mastodon-std.nix [hunk 14]
```diff
@@ -9,8 +9,8 @@ let
   '';
 
   hosts = ''
-    192.168.2.10 ca.local
-    192.168.2.11 mastodon.local
```
Comment (turion) [inline nixos/tests/web-apps/mastodon/mastodon-std.nix:13]: Are the ips of any particular importance? If not, why not keep the old ones and reuse them in the other test?

File: nixos/tests/web-apps/mastodon/mastodon-std.nix [hunk 15]
```diff
@@ -144,8 +142,12 @@ in
     server.wait_for_open_port(55000)
     server.wait_for_open_port(55001)
 
-    # Check Mastodon version from remote client
+    # Check that mastodon-media-auto-remove is scheduled
+    server.succeed("systemctl status mastodon-media-auto-remove.timer")
+
+    # Check work Mastodon from remote client
     client.succeed("curl --fail https://mastodon.local/api/v1/instance | jq -r '.version' | grep '${pkgs.mastodon.version}'")
+    client.succeed("curl --fail $(curl https://mastodon.local/api/v1/instance 2> /dev/null | jq -r .thumbnail) --output /dev/null")
```
Comment (turion) [inline nixos/tests/web-apps/mastodon/mastodon-std.nix:150]: This is good, but it's orthogonal to this PR. Maybe we can merge this, and all other changes that are unrelated to the remote db first in a separate PR? It would be easier because those are changes that everyone can easily agree upon, so it would be merged quicker.

File: nixos/tests/web-apps/mastodon/mastodon-ext.nix [hunk 16]
```diff
@@ -0,0 +1,253 @@
...
+          crt = "${test-certificates}/intermediate_ca.crt";
+          key = "${test-certificates}/intermediate_ca.key";
+          db = {
+            type = "badger";
+            dataSource = "/var/lib/step-ca/db";
+          };
+          authority = {
+            provisioners = [
+              {
+                type = "ACME";
```
Comment (turion) [inline nixos/tests/web-apps/mastodon/mastodon-ext.nix:49]: The CA node looks just like the CA node in the local-db test. We could remove this duplication by having a `generic.nix` test file that both tests import which contains all the joint settings. Then it's easier to see what actually changed.

File: nixos/tests/web-apps/mastodon/mastodon-ext.nix [hunk 17]
```diff
@@ -0,0 +1,253 @@
...
+        };
+        extraHosts = hosts;
+        firewall.allowedTCPPorts = [ 55001 55002 ];
+      };
+
+      services.redis.servers.mastodon = {
+        enable = true;
+        bind = "127.0.0.1";
+        port = 31637;
+      };
```
Comment (turion) [inline nixos/tests/web-apps/mastodon/mastodon-ext.nix:150]: Why configure it if it's the default setting anyways?

File: nixos/tests/web-apps/mastodon/mastodon-std.nix (hunk 14 above)
Comment (Izorkin) [reply nixos/tests/web-apps/mastodon/mastodon-std.nix]: It is easier to navigate the configuration by IP-addresses:
192.168.2.10x - databases and nginx
192.168.2.20x - mastodon service and client

File: nixos/tests/web-apps/mastodon/mastodon-ext.nix (hunk 16 above)
Comment (Izorkin) [reply nixos/tests/web-apps/mastodon/mastodon-ext.nix]: Removed CA service.

**Comment by Izorkin** [comment]
Split PR into https://github.com/NixOS/nixpkgs/pull/202398 and https://github.com/NixOS/nixpkgs/pull/202839

File: nixos/tests/web-apps/mastodon/mastodon-ext.nix (hunk 17 above)
Comment (Izorkin) [reply nixos/tests/web-apps/mastodon/mastodon-ext.nix]: Removed these lines.

File: nixos/tests/web-apps/mastodon/mastodon-std.nix (hunk 15 above)
Comment (Izorkin) [reply nixos/tests/web-apps/mastodon/mastodon-std.nix]: Moved changes to another PR.

File: nixos/doc/manual/release-notes/rl-2211.section.md [hunk 18]
```diff
@@ -340,6 +340,7 @@ Available as [services.patroni](options.html#opt-services.patroni.enable).
     "docker"
   ];
   ```
+- In `mastodon` it is now necessary to specify location of file with `PostgreSQL` database password. In `services.mastodon.database.passwordFile` parameter default value `/var/lib/mastodon/secrets/db-password` has been changed to `null`.
```
Comment (turion) [inline nixos/doc/manual/release-notes/rl-2211.section.md:343]: We have missed the merge window for that release, the notes will have to be moved to the next release.

File: nixos/doc/manual/release-notes/rl-2211.section.md [hunk 19]
```diff
@@ -359,6 +360,8 @@ Available as [services.patroni](options.html#opt-services.patroni.enable).
 
 - `mastodon` now automatically removes remote media attachments older than 30 days. This is configurable through `services.mastodon.mediaAutoRemove`.
 
+- `mastodon` now supports connection to a remote `PostrgeSQL` database.
```
Comment (turion) [inline nixos/doc/manual/release-notes/rl-2211.section.md:363]:
```suggestion
- `mastodon` now supports connection to a remote `PostgreSQL` database.
```

**Comment by turion** [comment]
See https://github.com/turion/nixpkgs/tree/update-mastodon-init-db on how you might extract the common part of the test scripts into a single file.

**Comment by Izorkin** [comment]
> See https://github.com/turion/nixpkgs/tree/update-mastodon-init-db on how you might extract the common part of the test scripts into a single file.

Thanks, added commit.

**Comment by turion** [comment]
I pushed a further commit to demonstrate how you can further share common config between the tests. I messed up something in the process, but maybe it serves as inspiration if you want to reduce further duplication.

**Comment by Izorkin** [comment]
> I pushed a further commit to demonstrate how you can further share common config between the tests. I messed up something in the process, but maybe it serves as inspiration if you want to reduce further duplication.

I think current version is optimal.

File: nixos/doc/manual/release-notes/rl-2211.section.md (hunk 18 above)
Comment (Izorkin) [reply nixos/doc/manual/release-notes/rl-2211.section.md]: Moved release notes.

File: nixos/doc/manual/release-notes/rl-2211.section.md (hunk 19 above)
Comment (Izorkin) [reply nixos/doc/manual/release-notes/rl-2211.section.md]: Fixed.

**Comment by Izorkin** [comment]
Rebased PR.

File: nixos/modules/services/web-apps/mastodon.nix (hunk 8 above)
Comment (erictapen) [reply nixos/modules/services/web-apps/mastodon.nix:557]: I also looked into it and it seems like there is no way to just provide a file with a password in it to postgresql. Only supported way seems to be the pgpassfile syntax. Let's leave it like this then :+1:

**Comment by Izorkin** [comment]
Rebased and resolve conflicts.

File: nixos/modules/services/web-apps/mastodon.nix (hunk 7 above)
Comment (erictapen) [reply nixos/modules/services/web-apps/mastodon.nix:492]: Yeah it's a bit more complicated to configure, On my services I do it via `services.postgresql.authentication`. Also the PostgreSQL instance doesn't necessarily run NixOS at all. I still think that this assertion assumes too much.

Comment (Izorkin) [reply nixos/modules/services/web-apps/mastodon.nix:492]: Delete these lines?

File: nixos/modules/services/web-apps/mastodon.nix [hunk 20]
```diff
@@ -37,6 +38,7 @@ let
 
     TRUSTED_PROXY_IP = cfg.trustedProxy;
   }
+  // (if (cfg.database.host != "/run/postgresql" && cfg.database.port != null) then { DB_PORT = toString cfg.database.port; } else {})
   // (if cfg.smtp.authenticate then { SMTP_LOGIN  = cfg.smtp.user; } else {})
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/mastodon.nix:42]: We should use optionalAttrs here

File: nixos/modules/services/web-apps/mastodon.nix [hunk 21]
```diff
@@ -512,7 +537,9 @@ in {
         OTP_SECRET="$(cat ${cfg.otpSecretFile})"
         VAPID_PRIVATE_KEY="$(cat ${cfg.vapidPrivateKeyFile})"
         VAPID_PUBLIC_KEY="$(cat ${cfg.vapidPublicKeyFile})"
+      '' + (if cfg.database.passwordFile != null then ''
         DB_PASS="$(cat ${cfg.database.passwordFile})"
+      '' else "") + ''
       '' + (if cfg.smtp.authenticate then ''
         SMTP_PASSWORD="$(cat ${cfg.smtp.passwordFile})"
       '' else "") + ''
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/mastodon.nix:545]: We should use lib.optionalString

File: nixos/modules/services/web-apps/mastodon.nix [hunk 22]
```diff
@@ -541,9 +577,18 @@ in {
...
+        rm $PGPASSFILE
+        unset PGPASSFILE
       '';
       path = [ cfg.package pkgs.postgresql ];
-      environment = env;
+      environment = env // (if (!databaseActuallyCreateLocally)
+        then {
+          PGHOST = cfg.database.host;
+          PGUSER = cfg.database.user;
+        }
```
Comment (SuperSandro2000) [inline nixos/modules/services/web-apps/mastodon.nix:589]: This can also use optionlAttr

File: nixos/modules/services/web-apps/mastodon.nix (hunk 7 above)
Comment (Izorkin) [reply nixos/modules/services/web-apps/mastodon.nix:492]: Removed.

File: nixos/modules/services/web-apps/mastodon.nix (hunk 21 above)
Comment (Izorkin) [reply nixos/modules/services/web-apps/mastodon.nix:545]: Fixed.

File: nixos/modules/services/web-apps/mastodon.nix (hunk 22 above)
Comment (Izorkin) [reply nixos/modules/services/web-apps/mastodon.nix:589]: Fixed.

File: nixos/modules/services/web-apps/mastodon.nix (hunk 20 above)
Comment (Izorkin) [reply nixos/modules/services/web-apps/mastodon.nix:42]: Thanks, fixed.

File: nixos/tests/web-apps/mastodon/script.nix [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+  server.succeed("mastodon-tootctl accounts create bob --email=bob@example.com")
+  server.succeed("mastodon-tootctl accounts approve bob")
+  server.succeed("mastodon-tootctl accounts delete bob")
+
+  # Manage IP access
+  server.succeed("mastodon-tootctl ip_blocks add 192.168.0.0/16 --severity=no_access")
+  server.succeed("mastodon-tootctl ip_blocks export' | grep '192.168.0.0/16")
+  server.fail("mastodon-tootctl ip_blocks export' | grep '172.16.0.0/16")
+  client.fail("curl --fail https://mastodon.local/about")
+  server.succeed(mastodon-tootctl ip_blocks remove 192.168.0.0/16")
```
Comment (erictapen) [inline nixos/tests/web-apps/mastodon/script.nix:47]:
```suggestion
  server.succeed("mastodon-tootctl ip_blocks remove 192.168.0.0/16")
```
This makes the test fail currently

Comment (Izorkin) [reply nixos/tests/web-apps/mastodon/script.nix:47]: Thanks, fixed.

File: nixos/modules/services/web-apps/mastodon.nix [hunk 24]
```diff
@@ -37,7 +38,8 @@ let
 
     TRUSTED_PROXY_IP = cfg.trustedProxy;
   }
-  // (if cfg.smtp.authenticate then { SMTP_LOGIN  = cfg.smtp.user; } else {})
+  // (if (cfg.database.host != "/run/postgresql" && cfg.database.port != null) then { DB_PORT = toString cfg.database.port; } else {})
```
Comment (erictapen) [inline nixos/modules/services/web-apps/mastodon.nix:41]: This also can be solved better by `lib.optionalAttrs`

Comment (Izorkin) [reply nixos/modules/services/web-apps/mastodon.nix]: Fixed!

**Review by erictapen — APPROVED** [review]
I feel like this is in good shape to merge now. If any issues would appear, they might better be solved in separate issues/PR

Big thanks to @Izorkin for your patience on this. 21 months is a long time for a PR, even on Nixpkgs. I'm glad that we finally have a working setup for remote postgresql and I'm especially glad about the additional test coverage here. The newly modularized testing infrastructure will prove it's use in the future, I'm sure.

I additionally checked, that both tests run through:

```
nix build .#nixosTests.mastodon.standard
nix build .#nixosTests.mastodon.remote-postgresql
```

**Comment by justinas** [comment]
@Izorkin Huge kudos for all your work contributing this, this is a big deal in allowing for more flexible Mastodon deployments on NixOS.

**Comment by f0x52** [comment]
These changes have made the Mastodon module much more opinionated than it used to be:
- [this assertion](https://github.com/NixOS/nixpkgs/blob/nixos-23.05/nixos/modules/services/web-apps/mastodon.nix#L570) prevents configuration with `createLocally = false` and `host = "/run/postgresql"`, even though that's a perfectly valid setup. Workaround for now is setting `host = "/var/run/postgresql"`
- https://github.com/NixOS/nixpkgs/blob/nixos-23.05/nixos/modules/services/web-apps/mastodon.nix#L657 forces the passwordFile to be set (to /dev/null I guess?) even though it's not needed, like when using peer auth
